internal/controller/controllers/agent_controller.go

```diff
@@ -80,69 +83,262 @@ func (r *AgentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	if err = r.Get(ctx, kubeKey, clusterDeployment); err != nil {
 		errMsg := fmt.Sprintf("failed to get clusterDeployment with name %s in namespace %s",
 			agent.Spec.ClusterDeploymentName.Name, agent.Spec.ClusterDeploymentName.Namespace)
-		Requeue = false
-		clientError := true
-		if !k8serrors.IsNotFound(err) {
-			Requeue = true
-			clientError = false
-		}
-		clusterDeploymentRefErr := newKubeAPIError(errors.Wrapf(err, errMsg), clientError)
 		// Update that we failed to retrieve the clusterDeployment
-		r.updateFailure(ctx, agent, clusterDeploymentRefErr)
-		return ctrl.Result{Requeue: Requeue}, nil
+		return r.updateStatus(ctx, agent, nil, errors.Wrapf(err, errMsg), !k8serrors.IsNotFound(err))
 	}
 
 	// Retrieve cluster for ClusterDeploymentName from the database
 	cluster, err := r.Installer.GetClusterByKubeKey(kubeKey)
 	if err != nil {
 		if gorm.IsRecordNotFoundError(err) {
-			Requeue = true
 			inventoryErr = common.NewApiError(http.StatusNotFound, err)
+			internalError = false
 		} else {
-			Requeue = false
 			inventoryErr = common.NewApiError(http.StatusInternalServerError, err)
+			internalError = true
 		}
 		// Update that we failed to retrieve the cluster from the database
-		r.updateFailure(ctx, agent, inventoryErr)
-		return ctrl.Result{Requeue: Requeue}, nil
+		return r.updateStatus(ctx, agent, nil, inventoryErr, internalError)
```

Comment:
maybe you can switch all this block into 
```suggestion
		return r.updateStatus(ctx, agent, nil, err, !gorm.IsRecordNotFoundError(err))
```

Reply:
Done